Purge Git history in Docker image (saves > 3 MiB)

bittner opened the pull request:
This change:

- reduces the Docker image size by more than 3 MiB using a [shallow Git clone](https://stackoverflow.com/questions/21833870/how-do-i-shallow-clone-a-repo-on-a-specific-branch) of the CRS repository,
- which also speeds up the Docker build significantly.
- removes one image layer and the software installed only for cloning the repo with Git,
- replaces the deprecated `MAINTAINER` directive by the [suggested](https://docs.docker.com/engine/reference/builder/#maintainer) `LABEL`

bittner commented:
The build currently fails due to issue #1344.

theMiddleBlue commented:
thanks @bittner !

can you rebase all commits in a single one?

bittner commented:
Squash the commits? Is this important?

With the single commits I demonstrate the different possibilities. This way it's obvious that there are 3 strategies involved to save space. With the second obsoleting the first one. This may be worth keeping?

Let me know! -- No problem otherwise.

theMiddleBlue commented:
oky no problem, just because It'd be good if they are merged so only 1 commit appears in the history

bittner commented:
Done. There's just a single commit now. :checkered_flag:

bittner commented:
BTW, I suspected that `iproute2` was used in the image for the CRS. That's why I'm not uninstalling it. If my assumption is wrong we can uninstall it either.

bittner commented:
And Python is actually used by the [Docker entrypoint](https://github.com/SpiderLabs/owasp-modsecurity-crs/blob/v3.2/dev/util/docker/docker-entrypoint.sh)! -- I probably need to only uninstall `git` and `ca-certificates`, can you confirm?

theMiddleBlue commented:
uhm... maybe it is required by the `docker-entrypoint.sh` but I'm not sure about that. Maybe @csanders-git can answer :)

bittner commented:
Anything I can do to help clarify the situation?

fgsch commented:
Now that the tests are fixed this needs to be updated.

Personally, I'd prefer if we keep git and ca-certificates in the image.
Is there any particular reason you want to remove them?

csanders-git commented:
This looks fine for right now -- We will be changing the location where CRS is extracted to soon -- but just not this second.

csanders-git commented:
Using the new base image and looks good!

bittner commented:
> Cloning into '/opt/owasp-modsecurity-crs-3.2'...
> warning: Could not find remote branch fe745bffc4713cb68cf53d79db5a9e069518fd47 to clone.
> fatal: Remote branch fe745bffc4713cb68cf53d79db5a9e069518fd47 not found in upstream origin

Hmmm ...

Local building works, though. :confused:

csanders-git commented:
you should be able to test it by just cloning your repo, i pushed it there ... github tried to do magic to remove conflicts and just made it worse -- which it does.

bittner commented:
Fine. So, is there anything left to do or clarify before this PR can get merged?

**A side note:** As soon as we've merged this, I would suggest to switch to a simpler branching model (e.g. GitHub Flow, only `master`, no long-living topic branches) and handle the CRS versions in the code (instead of via branches). I found it dangerous and confusing to have a main branch other than the usual `master` when doing the rebase.

fgsch commented:
Just wanted to reiterate my question above:

> Personally, I'd prefer if we keep git and ca-certificates in the image.
> Is there any particular reason you want to remove them?

fgsch commented:
Also tests are failing. Something is fubar with the origin branch:

```
warning: Could not find remote branch fe745bffc4713cb68cf53d79db5a9e069518fd47 to clone.
fatal: Remote branch fe745bffc4713cb68cf53d79db5a9e069518fd47 not found in upstream origin
```

bittner commented:
My reasoning about additional software in a security-related container image is: Additional software ...

- increases both the download size of an image, and
- increases the potential for attack vectors.

(Hardened systems should host only the bare minimum of software need to keep the system running, so intruders don't have tools at hand to continue their journey easily. This is most probably not the case of this `owasp/modsecurity:2.9-apache-ubuntu` derived image.)

<strike>`ca-certificates` is already included in the [base of the base image](https://github.com/CRS-support/modsecurity-docker/blob/v2/ubuntu-apache/Dockerfile#L10). We can safely drop adding it again, I guess. If I'm not totally mistaken.</strike> _That is a multi-stage build. Sorry!_ -- So, the fair question is: Are the CA certificates needed for running the container or tools therein?

Is `git` needed for the CRS or tools to work in the image? Or would it be exceptionally useful?

fgsch commented:
I understand the reasoning but those two should have low impact IMO and are useful.
I won't cry if they go though but they had come handy in the past.

bittner commented:
I'm not anyone to take decisions here. I'm willing to update the PR if it makes sense. No problem!

csanders-git commented:
i'll be jumping on this later today/this weekend -- promise, thanks for the patience.

dune73 commented:
@csanders-git ?

ghost commented:
@bittner if you use `git gc` it should compress the repo even more / substantially. Also can look into the `bfg` tool as well if you really want to get it down.

bittner commented:
I've removed all the purging now, hence only the shallow cloning takes place.

The `apt-get clean` seems to be unneeded as it happens in one of the base images (according to the [Docker documentation](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/)), so I omitted that, too.

The change is now reduced to the plain essential. I believe, this is now easy to merge. I'd be happy to see this happen soon. Thank you!

bittner commented:
Rebased the branch on top of `v3.2/dev`. Ready for merging.

fgsch commented:
LGTM

fgsch commented:
Actually, I take that back. `git clone -b` does not work with SHA1, so the proposed changes are not equivalent to the existing command.

bittner commented:
Okay, I see.

If it's really necessary to checkout a specific commit (as opposed to a branch) we need to proceed as until now, and purge the Git history in the end. I thought this would work with the suggested `git gc` from above, but it doesn't:

```bash
git gc --prune=$(git rev-parse HEAD)
```

So, to save space, my suggestion would be to simply remove the entire `.git` folder at the end. That's pragmatic. Simple and effective.

Does that sound like something we want?

bittner commented:
Fine like this?

fgsch commented:
Another possibility is:

```
mkdir /opt/owasp-modsecurity-crs-3.2 && \
cd /opt/owasp-modsecurity-crs-3.2 && \
git init && \
git remote add origin https://github.com/${REPO}.git && \
git fetch --depth 1 origin ${COMMIT} && \
git checkout FETCH_HEAD
```

(Ref. https://stackoverflow.com/questions/31278902/how-to-shallow-clone-a-specific-commit-with-depth-1).

You can also drop the extra copy and simply clone in /etc/apache2/modsecurity.d/owasp-crs.

bittner committed:
Make shallow clone of CRS repo in Docker image (saves > 3 MiB)

Combine image layers for cloning the CRS repo

Symlink to repo instead of creating identical copy

Replace deprecated MAINTAINER directive

bittner commented:
Alright, did as you said. I used a symlink, though. Makes sense, I guess.

Are we ready to merge now?

csanders-git commented:
@bittner , i'm gonna migrate your changes into the PR i'm working on -- for each of the new docker images -- take a look.

csanders-git commented:
Travis CI had a problem with this and honestly I don't quite know what the issue is. 

For instance in my PR's we had 
REPO - csanders-git/owasp-modsecurity-crs
COMMIT - b57355ad9f6e1dacce65ada5c3ff49a8a94e063c
which would result in: 
fatal: Couldn't find remote ref b57355ad9f6e1dacce65ada5c3ff49a8a94e063c

Now the weird part is that this only happened from travis -- I know there is some concern about allowing access by uploadpack.allowReachableSHA1InWant, and if Github supports it. In either event, I ended up adding the branch name temporarily on my version and fetch just the branch, then checkout the commit.

bittner commented:
My name is not @brittner. LOL  :smile:

csanders-git commented:
haha, not sure where i came up with that :-P

bittner commented:
I see you have integrated the provided changes in the v3.2/dev branch for `Dockerfile-2.9-apache` already (via https://github.com/SpiderLabs/owasp-modsecurity-crs/commit/fe660d4930c3ed632ae4ed26e2b46e4ef69b696a).

If you plan to integrate them manually also into `Dockerfile` we can easily close this PR. Otherwise ... I've rebased the PR again, so you can simply press the Merge button.

csanders-git commented:
So - the current thought is to remove the base (now broken) Dockerfile. Any objections @bittner ?

bittner commented:
Sure. For a production-grade container image we'd have to do a few things differently, anyway. As far as I can tell the current Dockerfile is really just (or mainly) for testing CRS rules.

csanders-git commented:
@bittner i've been thinking more and more about just merging the modsec and CRS images. Sure it'd take a long time to build, but it'd avoid having to maintain two different images. What are your thoughts i'm sure we'll talk about it at our meeting monday (@franbuehler)

bittner commented:
I would try to avoid that. ModSecurity and the CRS are two different projects, different from the technology used, and two separate repositories.

The build time is just one aspect. You actually lose flexibility (w.r.t. combining versions of the two). Manageability decreases for sure, it's just easier to understand a product when changes and code base in general are small.

csanders-git commented:
This was integrated into the docker rewrites. Thank you for the commit!
